File: plugins/notify/like.rb
# -*- coding: utf-8; -*-
#
# plugins/notify/like.rb : massr notify plugin for global like
#
# Copyright (C) 2013 by The wasam@s production
# https://github.com/wasamas/massr
#
# Distributed under GPL
#
# Usage:
#    "plugin": {
#       "notify/like party": {
#          "label": "Do you attend the party?",
#          "delete": "any"   // "any" or "owner"
#       }
#    }
#
module Massr
	module Plugin::Notify
		class Like
			SETTINGS = {} unless const_defined? :SETTINGS

			attr_reader :label, :delete

			def self.likes(id)
				JSON.parse(Massr::PluginSetting.get(id) || '{}')
			end

			def self.update(id, likes)
				Massr::PluginSetting.set(id, likes.to_json)
			end

			def self.cancel(id, name)
				tmp = likes(id)
				val = tmp.delete(name)
				update(id, tmp)
				return val
			end

			def self.add(id, me)
				tmp = likes(id)
				tmp[me.massr_id] = [me.twitter_icon_url, me.twitter_icon_url_https]
				update(id, tmp)
			end

			def self.delete(id, name, me)
				case SETTINGS[id].delete
				when 'owner'
					if name == me.massr_id
						return cancel(id, name)
					else
						return nil
					end
				when 'any'
					return cancel(id, name)
				end
				return nil
			end

			def self.to_json(id)
				likes(id).to_json
			end

			def initialize(plugin_id, opts)
				@id = plugin_id
				SETTINGS[@id] = self

				@label = opts['label'] || ''
				@delete = opts['delete'] || 'owner'
			end

			def render
				html = "#{@label}&nbsp;"
				html << %Q|<span class="pull-right"><a class="close" href="#" id="#{@id}-like"><img src="/img/wakaranaiwa.png"></a>|
				html << %Q|<a class="close" href="#" id="#{@id}-unlike"><img src="/img/wakaruwa.png"></a></span>|
				html << %Q|<span id="#{@id}">#{icons}</span>|
			end

		private

			def icons
				self.class.likes(@id).map{|user, (http, https)|
					case @delete
					when "any"
						%Q|<a href="#" class="#{@id}-delete"><img class="massr-icon-mini" src="#{https}" alt="#{user}" title="delete #{user}"></a>&nbsp;|
					else # "owner"
						%Q|<img class="massr-icon-mini" src="#{https}" alt="#{user}">|
					end
				}.join
			end
		end
	end

	class App < Sinatra::Base
		get '/plugin/notify/like/:plugin.json' do
			Massr::Plugin::Notify::Like.to_json(params[:plugin])
		end

		post '/plugin/notify/like/:plugin' do
			me = User.find_by(id: session[:user_id])
			Massr::Plugin::Notify::Like.add(params[:plugin], me)
			Massr::Plugin::Notify::Like.to_json(params[:plugin])
		end

		delete '/plugin/notify/like/:plugin/:user' do
			me = User.find_by(id: session[:user_id])
			Massr::Plugin::Notify::Like.delete(params[:plugin], params[:user], me)
			Massr::Plugin::Notify::Like.to_json(params[:plugin])
		end
	end
end

File: assets/js/plugins/notify/like.js
/*
 * notify/like.js : notify like plugin for massr
 *
 * Copyright (C) 2015 by The wasam@s production
 *
 * Distributed under GPL
 */
import Massr from '../../massr';

$(function(){
	var me = Massr.me;
	var _ = Massr.settings['local'];

	Massr.plugin_notify_like = function(id, opts){
		var del = opts['delete'] || 'owner';
		var myIcon = $('#'+id+' img[alt='+me+']').length !== 0;

		var likeButton = $('#'+id+'-like');
		var unlikeButton = $('#'+id+'-unlike');
		if(myIcon){
			likeButton.hide();
		}else{
			unlikeButton.hide();
		}

		likeButton.on('click', function(){
			$.ajax({
				url: '/plugin/notify/like/' + id,
				type: 'POST',
				dataType: 'json'
			}).done(function(result){
				plugin_notify_like_draw_icons(id, result);
			}).fail(function(XMLHttpRequest, textStatus, errorThrown){
				message.error('(' + textStatus + ')');
			});

			return false;
		});

		unlikeButton.on('click', function(){
			$.ajax({
				url: '/plugin/notify/like/' + id + '/' + me,
				type: 'DELETE',
				dataType: 'json'
			}).done(function(result){
				plugin_notify_like_draw_icons(id, result);
			}).fail(function(XMLHttpRequest, textStatus, errorThrown){
				message.error('(' + textStatus + ')');
			});

			return false;
		});

		$('#' + id).on('click', 'a.' + id + '-delete', function(){
			var name = $('img', this).attr('alt');

			$.ajax({
				url: '/plugin/notify/like/' + id + '/' + name,
				type: 'DELETE',
				dataType: 'json'
			}).done(function(result){
				plugin_notify_like_draw_icons(id, result);
			}).fail(function(XMLHttpRequest, textStatus, errorThrown){
				message.error('(' + textStatus + ')');
			});

			return false;
		});

		Massr.intervalFunctions.push(function(){
			$.getJSON('/plugin/notify/like/' + id + '.json', function(json){
				plugin_notify_like_draw_icons(id, json);
			});
		});

		function plugin_notify_like_draw_icons(id, icons){
			var elem = $('#' + id);

			elem.empty();
			$('#'+id+'-like').show();
			$('#'+id+'-unlike').hide();
			$.each(icons, function(name, val){
				if(del == 'any'){
					elem.append(
						$('<a>').addClass(id + '-delete').attr('href', '#').append(
							$('<img>').
							addClass('massr-icon-mini').
							attr('src', val[1]).
							attr('alt', name).
							attr('title', 'delete ' + name)
						)
					).append('&nbsp;')
				}else{
					elem.append(
						$('<img>').
						addClass('massr-icon-mini').
						attr('src', val[1]).
						attr('alt', name).
						attr('title', name)
					)
				}
				if(me == name){
					$('#'+id+'-like').hide();
					$('#'+id+'-unlike').show();
				}
			});
		}
	}
});
